### PR TITLE
Feat: Support partitioned_by expressions for bigquery

### PR DESCRIPTION
**Pull request by crericha**

### CODE REVIEW

File: sqlmesh/core/engine_adapter/spark.py [hunk 1]
```diff
@@ -247,7 +247,10 @@ def _create_table_properties(
         if partitioned_by:
             partition_columns_property = exp.PartitionedByProperty(
                 this=exp.Schema(
-                    expressions=[exp.to_identifier(column) for column in partitioned_by]
+                    expressions=[
+                        column.this if isinstance(column, exp.Column) else column
```
Comment (izeigerman) [inline sqlmesh/core/engine_adapter/spark.py:251]: Should we fail if we detect a non-column expression?

File: sqlmesh/core/model/definition.py [hunk 2]
```diff
@@ -1644,12 +1648,16 @@ def _single_value_or_tuple(values: t.Sequence) -> exp.Identifier | exp.Tuple:
     )
 
 
+def _single_expr_or_tuple(values: t.Sequence) -> exp.Expression | exp.Tuple:
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1651]: [Nit] `t.Sequence[exp.Expression]`?

File: sqlmesh/core/model/meta.py [hunk 3]
```diff
@@ -25,15 +24,21 @@
...
     """
 
+    YEAR = "year"
+    MONTH = "month"
     DAY = "day"
     HOUR = "hour"
     MINUTE = "minute"
 
+    @property
+    def is_date_type(self) -> bool:
```
Comment (izeigerman) [inline sqlmesh/core/model/meta.py:39]: [Nit] To me it's more of a "granularity" than type.

File: sqlmesh/core/model/definition.py [hunk 4]
```diff
@@ -557,7 +557,11 @@ def validate_definition(self) -> None:
             ConfigError
         """
         if self.partitioned_by:
-            unique_partition_keys = {k.strip().lower() for k in self.partitioned_by}
+            unique_partition_keys = {
+                col.name.strip().lower()
+                for expr in self.partitioned_by
+                for col in expr.find_all(exp.Column)
```
[the remark this answers is not in the document]
Comment (izeigerman) [reply sqlmesh/core/model/definition.py:563]: Found this logic in the column validator.

File: sqlmesh/dbt/model.py [hunk 5]
```diff
@@ -98,6 +101,14 @@ def _validate_list(cls, v: t.Union[str, t.List[str]]) -> t.List[str]:
     def _validate_sql(cls, v: t.Union[str, SqlStr]) -> SqlStr:
         return SqlStr(v)
 
+    @validator("partition_by", pre=True)
+    def _validate_partition_by(cls, v: t.Dict[str, t.Any]) -> t.Dict[str, t.Any]:
+        if not v.get("field"):
+            raise ConfigError("'field' key required for partition_by.")
```
Comment (izeigerman) [inline sqlmesh/dbt/model.py:107]: This only true for BigQuery. For example in Spark you can just specify a list of strings
https://docs.getdbt.com/reference/resource-configs/spark-configs
```
{{ config(
    materialized='incremental',
    partition_by=['date_day'],
    file_format='parquet'
) }}
```

File: tests/core/test_model.py [hunk 6]
```diff
@@ -194,14 +193,22 @@ def test_model_validation_union_query():
...
 
-def test_partitioned_by():
-    expressions = parse(
-        """
+@pytest.mark.parametrize(
+    "partition_by_input, partition_by_output",
+    [
+        ("a", ["a"]),
+        ("(a, b)", ["a", "b"]),
+        ("TIMESTAMP_TRUNC(a, DAY)", ["TIMESTAMP_TRUNC(a, DAY)"]),
```
Comment (izeigerman) [inline tests/core/test_model.py:201]: Do we have a test for expressions that don't reference a column?

File: sqlmesh/core/engine_adapter/spark.py (hunk 1 above)
Comment (crericha) [reply sqlmesh/core/engine_adapter/spark.py:251]: If only column expressions are valid, then yes. Please confirm.

File: sqlmesh/core/model/meta.py [hunk 7]
```diff
@@ -136,6 +141,39 @@ def _cron_validator(cls, v: t.Any) -> t.Optional[str]:
...
+        elif isinstance(v, exp.Expression):
+            partitions = [v]
+        else:
+            dialect = values.get("dialect")
+            partitions = [
+                d.parse_one(entry, dialect=dialect) if isinstance(entry, str) else entry
+                for entry in ensure_list(v)
+            ]
+        partitions = [
+            exp.to_column(expr.name) if isinstance(expr, exp.Identifier) else expr
```
[the remark this answers is not in the document]
Comment (crericha) [reply sqlmesh/core/model/meta.py:160]: The parser returns an identifier if a single value. I took a more cautious approach here. I could potentially only convert if a single value.

File: sqlmesh/dbt/model.py (hunk 5 above)
Comment (crericha) [reply sqlmesh/dbt/model.py:107]: Thanks for pointing out. I didn't realize dbt spark had the same field with a different format.

File: tests/core/test_model.py (hunk 6 above)
Comment (crericha) [reply tests/core/test_model.py:201]: No, but we should. Good catch.

File: sqlmesh/core/engine_adapter/spark.py (hunk 1 above)
Comment (izeigerman) [reply sqlmesh/core/engine_adapter/spark.py:251]: For vanilla spark - yes, I think so

File: sqlmesh/core/engine_adapter/spark.py [hunk 8]
```diff
@@ -237,18 +237,21 @@ def create_view(
...
+            for expr in partitioned_by:
+                if not isinstance(expr, exp.Column):
+                    raise SQLMeshError(
+                        f"PARTITIONED BY contains non-column value '{expr.sql(dialect='spark')}'."
+                    )
             partition_columns_property = exp.PartitionedByProperty(
-                this=exp.Schema(
-                    expressions=[exp.to_identifier(column) for column in partitioned_by]
-                ),
+                this=exp.Schema(expressions=partitioned_by),
```
Comment (izeigerman) [inline sqlmesh/core/engine_adapter/spark.py:254]: Is it ok that that this is a list of `Column` vs list of identifiers like it used to be?

**Review by izeigerman — APPROVED** [review]

File: sqlmesh/core/engine_adapter/spark.py (hunk 8 above)
Comment (crericha) [reply sqlmesh/core/engine_adapter/spark.py:254]: yes, verified via unit tests.

**Pull request merged**